From 475f187d9dcd20f269579a73db9350b63192cd5e Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Sat, 29 Oct 2022 19:06:34 +0800
Subject: [PATCH] Async entity ticking


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 107d80a1e..1b2f21cb2 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1520,7 +1520,7 @@ public abstract class World implements IBlockAccess {
         timings.entityTick.startTiming(); // Spigot
         // CraftBukkit start - Use field for loop variable
         co.aikar.timings.TimingHistory.entityTicks += this.entityList.size(); // Paper
-        worker.submit(AsyncConcurrentTraverseTask.wrapNewTask(this.entityList,entity->{
+        worker.execute(AsyncConcurrentTraverseTask.wrapNewTask(this.entityList,entity->{
             Entity entity1 = entity.bJ();
 
             if (entity1 != null) {
@@ -1568,7 +1568,7 @@ public abstract class World implements IBlockAccess {
             }
 
             this.methodProfiler.b();
-        },worker)).join();
+        },worker));
 
         timings.entityTick.stopTiming(); // Spigot
         this.methodProfiler.c("blockEntities");
@@ -1702,11 +1702,7 @@ public abstract class World implements IBlockAccess {
         if (this.O) {
             this.b.addAll(collection);
         } else {
-            Iterator iterator = collection.iterator();
-
-            while (iterator.hasNext()) {
-                TileEntity tileentity = (TileEntity) iterator.next();
-
+            for (TileEntity tileentity : collection) {
                 this.a(tileentity);
             }
         }
-- 
2.34.1

