From 6599d6cf8f02be80e9a4aeaac83f804ae0677b33 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Sat, 29 Oct 2022 15:46:00 +0800
Subject: [PATCH] Parallel entity ticking


diff --git a/src/main/java/co/aikar/util/Counter.java b/src/main/java/co/aikar/util/Counter.java
index 23ac07f2..f225d631 100644
--- a/src/main/java/co/aikar/util/Counter.java
+++ b/src/main/java/co/aikar/util/Counter.java
@@ -1,12 +1,11 @@
 package co.aikar.util;
 
 import com.google.common.collect.ForwardingMap;
-
-import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class Counter <T> extends ForwardingMap<T, Long> {
-    private final Map<T, Long> counts = new HashMap<>();
+    private final Map<T, Long> counts = new ConcurrentHashMap<>(); //M2ke4U -- HashMap -> ConcurrentHashMap
 
     public long decrement(T key) {
         return increment(key, -1);
-- 
2.34.1

